proposals/0000-amending-mrp.rst

```diff
@@ -0,0 +1,235 @@
+Example file:
+::
+  {-#LANGUAGE DeriveAnyClass#-}
+
+  newtype Id a = MkId a deriving (Functor, Applicative)
+  newtype Id2 a = MkId2 a deriving (Functor, Applicative)
+
+  instance Monad Id where
+    return = MkId
+    (>>) _ b = b
+    (>>=) = undefined
+
+  instance Monad Id2 where
+    return = pure
+    (>>) = (*>)
+    (>>=) = undefined
+
+Current warnings:
+::
+  example.hs:6:3: warning: [-Wnoncanonical-monad-instances]
+      Noncanonical ‘return’ definition detected
+      in the instance declaration for ‘Monad Id’.
+      ‘return’ will eventually be removed in favour of ‘pure’
+      Either remove definition for ‘return’ (recommended) or define as ‘return = pure’
+      See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
+    |
+  6 |   return = MkId
+    |   ^^^^^^^^^^^^^
+
+  example.hs:7:3: warning: [-Wnoncanonical-monad-instances]
+      Noncanonical ‘(>>)’ definition detected
+      in the instance declaration for ‘Monad Id’.
+      ‘(>>)’ will eventually be removed in favour of ‘(*>)’
+      Either remove definition for ‘(>>)’ (recommended) or define as ‘(>>) = (*>)’
+      See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
+    |
+  7 |   (>>) _ b = b
+    |   ^^^^^^^^^^^^
+
+Eventual warnings and errors (expected):
+::
+  example.hs:7:5: error: [-Wnoncanonical-monad-instances]
+      Noncanonical ‘return’ definition detected
+      in the instance declaration for ‘Monad Id’.
+      ‘return’ will eventually be removed in favour of ‘pure’
+      Remove definition for ‘return’ (recommended) or define as ‘return = pure’
+      See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
+    |
+  7 |     return = MkId
+    |     ^^^^^^^^^^^^^
+
+  example.hs:8:5: error: [-Wnoncanonical-monad-instances]
+      Noncanonical ‘(>>)’ definition detected
+      in the instance declaration for ‘Monad Id’.
+      ‘(>>)’ will eventually be removed in favour of ‘(*>)’
+      Remove definition for ‘(>>)’ (recommended) or define as ‘(>>) = (*>)’
+      See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
+    |
+  8 |     (>>) _ b = b
+    |     ^^^^^^^^^^^^
+
+  example.hs:12:5: error: [-Wredundant-canonical-monad-instances]
+      ‘return’ definition detected
+      in the instance declaration for ‘Monad Id’.
+      ‘return’ will eventually be removed in favour of ‘pure’
+      Remove definition for ‘return’
+      See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
+    |
+  12|     return = pure
+    |     ^^^^^^^^^^^^^
+
+  example.hs:13:5: error: [-Wredundant_canonical-monad-instances]
+      ‘(>>)’ definition detected
+      in the instance declaration for ‘Monad Id’.
+      ‘(>>)’ will eventually be removed in favour of ‘(*>)’
+      Remove definition for ‘(>>)’
+      See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
+    |
+  13|     (>>) = (*>)
+    |     ^^^^^^^^^^^
```

Comment:
Is it a good idea to leave the old warnings unchanged? I see that it recommends using `return = pure` but why not offer it as the only solution, now that removing it just emits another warning?

Reply:
not having a definition for `return` is the preferred way we want things to go in future. I guess we would prefer to tell them to only remove the definition, but for transparency I wanted to suggest either approach, and then let the other warning warn them off if they're still committed to defining every method.

Reply:
Personally I'd lean towards changing the message if the new warning is enabled by default, but leaving it as-is otherwise.

Reply:
I guess we probably shouldn't recommend people do something that we warn against, so I'll adjust the wording.